Fix confmap to preserve original type when a slice of a map is unmarshalled

samuelarogbonlo opened the pull request:
#### Description

This PR fixes an issue where the confmap package fails to preserve the original type when a slice of a map is unmarshalled. This occurs specifically when using environment variables with numeric values (e.g., SPLUNK_ACCESS_TOKEN=12345) in configurations with slices of maps.
The error manifests as:
`'headers[0].default_value' expected type 'string', got unconvertible type 'int', value: '12345'`

The root cause is that while the confmap package correctly handles type conversion for regular maps, it doesn't process maps within slices. I've added a new hook function (sliceOfMapsHookFunc) that specifically converts numeric values to strings within maps that are nested in slices during the unmarshalling process.

#### Link to tracking issue
Fixes #12793

#### Testing

I've added a unit test TestSliceOfMapsUnmarshal which verifies that numeric values in maps within slices are correctly converted to strings during unmarshalling. The test creates a configuration that mimics the scenario described in the issue and confirms that the unmarshalling process works as expected.


#### Documentation 

No documentation changes were needed as this is a bug fix for existing functionality.

samuelarogbonlo commented:
> Thanks for the PR!
> 
> I don't think this is the right way to solve the problem mentioned on the linked issue. I believe the problem stems from the use of `*configopaque.String` [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c1bef8e75ee1da4f65830de9e9389d7e24930e27/extension/headerssetterextension/config.go#L29), which we are not supporting properly.
> 
> I am against any solution that involves converting numeric values to strings indiscriminately: this leads to buggy and hard to understand behavior where something like `0123` ends up setting the string `"83"`. Same goes for any other type with multiple possible representations.

Thanks for the feedback! I understand your concerns about indiscriminately converting numeric values to strings, which could indeed lead to unexpected behavior.

I see that the issue is specifically related to `configopaque.String` rather than a general problem with maps in slices. I'll revise my approach to create a targeted hook specifically for `configopaque.String` types that preserves the original representation of values.

I'll work on implementing this more focused solution and update the PR soon.

mx-psi commented:
Awesome! See my comments over at #12793 to help you get started

samuelarogbonlo commented:
I've implemented the solution suggested by @mx-psi by adding pointer type handling to isStringyStructure and useExpandValue. This should properly address the issue with configopaque.String types in a more targeted way. The existing tests are now passing, showing that the fix works correctly.

mx-psi commented on test/confmap/test_program.go:
This should be in https://github.com/open-telemetry/opentelemetry-collector/tree/main/confmap/internal/e2e instead

mx-psi commented on test/confmap/test_program.go:
```diff
@@ -0,0 +1,87 @@
+	Action       string `mapstructure:"action"`
+	Key          string `mapstructure:"key"`
+	FromContext  string `mapstructure:"from_context"`
+	DefaultValue string `mapstructure:"default_value"`
```
This is not testing the issue fully. I believe the issue happens because of the type being `*configopaque.String`

samuelarogbonlo commented:
I've implemented a more targeted solution as suggested by @mx-psi. Instead of broadly converting all numeric values in maps within slices, this fix specifically addresses the issue with configopaque.String types by adding a dedicated hook that converts numeric values to strings when the target type is configopaque.String (either direct or pointer). I've added a test that verifies this fix correctly handles the scenario from the issue.

mx-psi commented on confmap/confmap.go:
```diff
@@ -219,13 +219,29 @@ func (l *Conf) ToStringMap() map[string]any {
+	// Add a specific hook for configopaque.String
+	configOpaqueHook := func(_ reflect.Type, t reflect.Type, data interface{}) (interface{}, error) {
+		// Check for both direct configopaque.String and pointer to it
+		isConfigOpaque := t.String() == "configopaque.String" ||
```
I think the solution if possible should work for any type whose underlying kind is `string`, that is, any type defined as `type T string`. I wonder if #12927 is enough to solve your issue. One way to test this would be to add the end to end test you created on top of the PR. If it works, I would be happy to first accept a PR just for the test

samuelarogbonlo replied:
Thank you pointing me to PR #12927. I've checked it out and verified that it does solve the issue with *configopaque.String fields. I applied the changes from that PR and ran my test case, which is now passing without needing any specific handling for configopaque.String.
As you suggested, would you prefer I just contribute my test case to help verify that PR #12927 properly addresses this specific scenario? I'd be happy to create a PR that just adds this test on top of #12927.

mx-psi replied:
Yeah, that would be it, let's wait until #12927 is merged and you can open a PR then. Apologies for the back-and-forth, but I think we arrived at a simple solution in the end!
